### PR TITLE
Spill in explain analyze, query summary and web UI

### PR DESCRIPTION
**Pull request by ilfrin**
This pull request adds optional information about spilled data volume in a couple contexts, including raw JSON, explain analyze output, query summary and the web UI. Not to clutter the already packed views, if there is no spilled data the "Spilled data" label is not included.

### CODE REVIEW

**Comment by ilfrin** [comment]
@martint not sure who's the best assignee to review this, suggestions?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java [hunk 1]
```diff
@@ -217,6 +222,8 @@ public QueryStats(
         checkArgument(outputPositions >= 0, "outputPositions is negative");
         this.outputPositions = outputPositions;
         this.operatorSummaries = ImmutableList.copyOf(requireNonNull(operatorSummaries, "operatorSummaries is null"));
+
+        this.spilledDataSize = spilledDataSize;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: require not null

File: presto-main/src/main/java/com/facebook/presto/execution/StageStats.java [hunk 2]
```diff
@@ -190,6 +195,8 @@ public StageStats(
         checkArgument(outputPositions >= 0, "outputPositions is negative");
         this.outputPositions = outputPositions;
         this.operatorSummaries = ImmutableList.copyOf(requireNonNull(operatorSummaries, "operatorSummaries is null"));
+
+        this.spilledDataSize = spilledDataSize;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStats.java]: require not null

File: presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java [hunk 3]
```diff
@@ -159,6 +165,8 @@ public DriverStats(
         this.outputPositions = outputPositions;
 
         this.operatorStats = ImmutableList.copyOf(requireNonNull(operatorStats, "operatorStats is null"));
+
+        this.spilledDataSize = spilledDataSize;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java]: require not null

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 4]
```diff
@@ -142,6 +146,8 @@ public OperatorStats(
         this.systemMemoryReservation = requireNonNull(systemMemoryReservation, "systemMemoryReservation is null");
         this.blockedReason = blockedReason;
 
+        this.spilledDataSize = spilledDataSize;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: require not null

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java [hunk 5]
```diff
@@ -168,6 +171,8 @@ public PipelineStats(
 
         this.operatorSummaries = ImmutableList.copyOf(requireNonNull(operatorSummaries, "operatorSummaries is null"));
         this.drivers = ImmutableList.copyOf(requireNonNull(drivers, "drivers is null"));
+
+        this.spilledDataSize = spilledDataSize;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java]: require not null

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java [hunk 6]
```diff
@@ -191,6 +196,8 @@ public TaskStats(
         this.outputPositions = outputPositions;
 
         this.pipelines = ImmutableList.copyOf(requireNonNull(pipelines, "pipelines is null"));
+
+        this.spilledDataSize = spilledDataSize;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: require not null

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 7]
```diff
@@ -327,6 +327,8 @@ public QueryInfo getQueryInfo(Optional<StageInfo> rootStage)
         boolean fullyBlocked = rootStage.isPresent();
         Set<BlockedReason> blockedReasons = new HashSet<>();
 
+        long spilledBytes = 0;
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Please capitalize first word in all your commit titles

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java [hunk 8]
```diff
@@ -164,7 +167,9 @@ public QueryStats(
             @JsonProperty("outputDataSize") DataSize outputDataSize,
             @JsonProperty("outputPositions") long outputPositions,
 
-            @JsonProperty("operatorSummaries") List<OperatorStats> operatorSummaries)
+            @JsonProperty("operatorSummaries") List<OperatorStats> operatorSummaries,
+
+            @JsonProperty("spilledDataSize") DataSize spilledDataSize)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: Move this to after outputPositions

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java [hunk 9]
```diff
@@ -437,8 +444,14 @@ public long getOutputPositions()
     public OptionalDouble getProgressPercentage()
     {
         if (!scheduled || totalDrivers == 0) {
-             return OptionalDouble.empty();
+            return OptionalDouble.empty();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: Revert, fix in separate commit if you want to.

File: presto-main/src/main/java/com/facebook/presto/execution/StageStats.java [hunk 10]
```diff
@@ -143,7 +146,9 @@ public StageStats(
             @JsonProperty("bufferedDataSize") DataSize bufferedDataSize,
             @JsonProperty("outputDataSize") DataSize outputDataSize,
             @JsonProperty("outputPositions") long outputPositions,
-            @JsonProperty("operatorSummaries") List<OperatorStats> operatorSummaries)
+            @JsonProperty("operatorSummaries") List<OperatorStats> operatorSummaries,
+            @JsonProperty("spilledDataSize") DataSize spilledDataSize)
+
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStats.java]: Newline?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java (hunk 1 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: Please address this

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java (hunk 8 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: Do this at every level of stat

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java [hunk 11]
```diff
@@ -142,7 +145,9 @@ public TaskStats(
             @JsonProperty("outputDataSize") DataSize outputDataSize,
             @JsonProperty("outputPositions") long outputPositions,
 
-            @JsonProperty("pipelines") List<PipelineStats> pipelines)
+            @JsonProperty("pipelines") List<PipelineStats> pipelines,
+
+            @JsonProperty("spilledDataSize") DataSize spilledDataSize)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: Move as discussed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStatsSummarizer.java [hunk 12]
```diff
@@ -128,14 +130,16 @@ private PlanNodeStatsSummarizer() {}
...
 
-                // An "internal" pipeline like a hash build, links to another pipeline which is the actual output for this plan node
-                if (operatorStats.getPlanNodeId().equals(outputPlanNode) && !pipelineStats.isOutputPipeline()) {
-                    continue;
-                }
                 if (processedNodes.contains(planNodeId)) {
                     continue;
                 }
 
+                planNodeSpilledBytes.merge(planNodeId, operatorStats.getSpilledDataSize().toBytes(), Long::sum);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStatsSummarizer.java]: Move down to be with the other merges

File: presto-main/src/main/resources/com/facebook/presto/server/plan.html [hunk 13]
```diff
@@ -181,6 +181,7 @@
                 ) +
                 "<div>Memory: " + stats.totalMemoryReservation + "</div>" +
                 "<div>Splits: Q:" + stats.queuedDrivers + ", R:" + stats.runningDrivers + ", F:" + stats.completedDrivers + "</div>" +
+                "<div>Spilled: " + stats.spilledDataSize + "</div>" +
```
Comment (raghavsethi) [inline presto-main/src/main/resources/com/facebook/presto/server/plan.html]: This real estate seems a little too valuable to add this.

**Comment by raghavsethi** [comment]
Also, please rebase prior to next review.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStatsSummarizer.java (hunk 12 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStatsSummarizer.java]: but that moves the merge below the if-continue clause, I'd rather keep this here

File: presto-main/src/main/resources/com/facebook/presto/server/plan.html (hunk 13 above)
Comment (ilfrin) [inline presto-main/src/main/resources/com/facebook/presto/server/plan.html]: What do you suggest then? I think not displaying this at all is not a good option either.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 7 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: done

**Comment by ilfrin** [comment]
@raghavsethi AC and rebased. Not applied two comments, please answer. Thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 14]
```diff
@@ -399,7 +399,10 @@ public OperatorStats getOperatorStats()
                 succinctBytes(memoryReservation.get()),
                 succinctBytes(systemMemoryContext.getReservedBytes()),
                 memoryFuture.get().isDone() ? Optional.empty() : Optional.of(WAITING_FOR_MEMORY),
-                info);
+
+                info
+
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: Revert - unrelated change

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 15]
```diff
@@ -92,7 +94,7 @@ public OperatorStats(
             @JsonProperty("outputDataSize") DataSize outputDataSize,
             @JsonProperty("outputPositions") long outputPositions,
 
-            @JsonProperty("blockedWall") Duration blockedWall,
+            @JsonProperty("spilledDataSize") DataSize spilledDataSize, @JsonProperty("blockedWall") Duration blockedWall,
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Fix

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 16]
```diff
@@ -417,7 +430,9 @@ public OperatorStats add(Iterable<OperatorStats> operators)
                 succinctBytes(systemMemoryReservation),
                 blockedReason,
 
-                (OperatorInfo) base);
+                (OperatorInfo) base
+
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Revert

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 17]
```diff
@@ -457,14 +472,15 @@ public OperatorStats summarize()
...
+                spilledDataSize, blockedWall,
                 finishCalls,
                 finishWall,
                 finishCpu,
                 finishUser,
                 memoryReservation,
                 systemMemoryReservation,
                 blockedReason,
-                (info != null && info.isFinal()) ? info : null);
+                (info != null && info.isFinal()) ? info : null
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Revert

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java [hunk 18]
```diff
@@ -459,8 +461,10 @@ public PipelineStats getPipelineStats()
                 succinctBytes(outputDataSize),
                 outputPositions,
 
-                ImmutableList.copyOf(operatorSummaries.values()),
-                drivers);
+                succinctBytes(spilledBytes.get()), ImmutableList.copyOf(operatorSummaries.values()),
+                drivers
+
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java]: Revert

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java [hunk 19]
```diff
@@ -384,9 +396,10 @@ public PipelineStats summarize()
                 processedInputPositions,
                 outputDataSize,
                 outputPositions,
-                operatorSummaries.stream()
+                spilledDataSize, operatorSummaries.stream()
                         .map(OperatorStats::summarize)
                         .collect(Collectors.toList()),
-                ImmutableList.of());
+                ImmutableList.of()
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java]: Revert

File: presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java [hunk 20]
```diff
@@ -390,6 +393,7 @@ public TaskStats getTaskStats()
                 processedInputPositions,
                 succinctBytes(outputDataSize),
                 outputPositions,
-                pipelineStats);
+                succinctBytes(spilledBytes.get()), pipelineStats
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: Fix

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java [hunk 21]
```diff
@@ -102,7 +104,9 @@ public TaskStats(DateTime createTime, DateTime endTime)
                 0,
                 new DataSize(0, BYTE),
                 0,
-                ImmutableList.of());
+                new DataSize(0, BYTE),
+                ImmutableList.of()
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: Revert

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java [hunk 22]
```diff
@@ -436,8 +451,9 @@ public TaskStats summarizeFinal()
                 processedInputPositions,
                 outputDataSize,
                 outputPositions,
-                pipelines.stream()
+                spilledDataSize, pipelines.stream()
                         .map(PipelineStats::summarize)
-                        .collect(Collectors.toList()));
+                        .collect(Collectors.toList())
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: Move end to same line

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java [hunk 23]
```diff
@@ -70,7 +70,8 @@
 
             new DataSize(28, BYTE),
             29,
-            ImmutableList.of());
+            new DataSize(30, BYTE), ImmutableList.of()
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java]: Fix

**Comment by ilfrin** [comment]
rebased

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 24]
```diff
@@ -365,6 +367,7 @@ public QueryInfo getQueryInfo(Optional<StageInfo> rootStage)
             }
             completeInfo = completeInfo && stageInfo.isCompleteInfo();
             operatorStatsSummary.addAll(stageInfo.getStageStats().getOperatorSummaries());
+            spilledBytes += stageStats.getSpilledDataSize().toBytes();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Move this to somewhere around L350 (so that it's close to other similar stats) or 368 (right after outputPositions)

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java [hunk 25]
```diff
@@ -452,4 +460,10 @@ public OptionalDouble getProgressPercentage()
         }
         return OptionalDouble.of(min(100, (completedDrivers * 100.0) / totalDrivers));
     }
+
+    @JsonProperty
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: We keep getters in the same order as the member variables, please put this after `getOutputPositions`

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java [hunk 26]
```diff
@@ -221,6 +226,9 @@ public QueryStats(
         this.outputDataSize = requireNonNull(outputDataSize, "outputDataSize is null");
         checkArgument(outputPositions >= 0, "outputPositions is negative");
         this.outputPositions = outputPositions;
+
+        this.spilledDataSize = requireNonNull(spilledDataSize);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: add requireNonNull message

File: presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java [hunk 27]
```diff
@@ -285,6 +287,7 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier, Su
                     operatorToStats.compute(id, (k, v) -> v == null ? operatorStats : v.add(operatorStats));
                 }
             }
+            spilledBytes += taskStats.getSpilledDataSize().toBytes();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStateMachine.java]: Move to after outputPositions

File: presto-main/src/main/java/com/facebook/presto/execution/StageStats.java [hunk 28]
```diff
@@ -194,6 +201,9 @@ public StageStats(
         this.outputDataSize = requireNonNull(outputDataSize, "outputDataSize is null");
         checkArgument(outputPositions >= 0, "outputPositions is negative");
         this.outputPositions = outputPositions;
+
+        this.spilledDataSize = requireNonNull(spilledDataSize);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStats.java]: requireNonNull message

File: presto-main/src/main/java/com/facebook/presto/execution/StageStats.java [hunk 29]
```diff
@@ -370,4 +380,10 @@ public long getOutputPositions()
     {
         return operatorSummaries;
     }
+
+    @JsonProperty
+    public DataSize getSpilledDataSize()
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/StageStats.java]: Same comment

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java [hunk 30]
```diff
@@ -204,6 +205,7 @@ public void transferMemoryToTaskContext(long bytes)
 
     public ListenableFuture<?> reserveSpill(long bytes)
     {
+        spilledBytes.getAndAdd(bytes);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: Do we actually need driver-level spill stats, or are pipeline level enough?

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: Also, why aren't we reducing this value in `freeSpill`?

File: presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java [hunk 31]
```diff
@@ -294,4 +302,10 @@ public long getOutputPositions()
     {
         return operatorStats;
     }
+
+    @JsonProperty
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverStats.java]: Same comment

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 32]
```diff
@@ -518,6 +521,7 @@ public void updateBytes(long bytes)
         {
             if (bytes > 0) {
                 driverContext.reserveSpill(bytes);
+                spilledBytes += bytes;
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: What happens if bytes < 0?

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 33]
```diff
@@ -131,6 +135,8 @@ public OperatorStats(
         checkArgument(outputPositions >= 0, "outputPositions is negative");
         this.outputPositions = outputPositions;
 
+        this.spilledDataSize = requireNonNull(spilledDataSize);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: requireNonNull message

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 34]
```diff
@@ -301,6 +307,12 @@ public DataSize getSystemMemoryReservation()
         return blockedReason;
     }
 
+    @JsonProperty
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Same comment

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 35]
```diff
@@ -342,6 +354,7 @@ public OperatorStats add(Iterable<OperatorStats> operators)
         long memoryReservation = this.memoryReservation.toBytes();
         long systemMemoryReservation = this.systemMemoryReservation.toBytes();
         Optional<BlockedReason> blockedReason = this.blockedReason;
+        long spilledBytes = this.spilledDataSize.toBytes();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Move next to outputPositions

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 36]
```diff
@@ -377,6 +390,8 @@ public OperatorStats add(Iterable<OperatorStats> operators)
                 blockedReason = operator.getBlockedReason();
             }
 
+            spilledBytes += operator.getSpilledDataSize().toBytes();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Move next to outputPositions

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java [hunk 37]
```diff
@@ -233,6 +234,7 @@ public void transferMemoryToTaskContext(long bytes)
 
     public synchronized ListenableFuture<?> reserveSpill(long bytes)
     {
+        spilledBytes.getAndAdd(bytes);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineContext.java]: Same comment - where is the change in freeSpill?

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java [hunk 38]
```diff
@@ -170,6 +174,8 @@ public PipelineStats(
         checkArgument(outputPositions >= 0, "outputPositions is negative");
         this.outputPositions = outputPositions;
 
+        this.spilledDataSize = requireNonNull(spilledDataSize);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java]: Same

File: presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java [hunk 39]
```diff
@@ -363,6 +369,12 @@ public long getOutputPositions()
         return drivers;
     }
 
+    @JsonProperty
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/PipelineStats.java]: Same

File: presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java [hunk 40]
```diff
@@ -168,6 +170,7 @@ public TaskState getState()
     public synchronized ListenableFuture<?> reserveSpill(long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
+        spilledBytes.getAndAdd(bytes);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java]: Same

File: presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java [hunk 41]
```diff
@@ -383,6 +390,12 @@ public int getRunningPartitionedDrivers()
         return runningPartitionedDrivers;
     }
 
+    @JsonProperty
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/TaskStats.java]: Same

**Comment by raghavsethi** [comment]
Let me know when comments are addressed.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 7 above)
Comment (ilfrin) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: done

**Comment by ilfrin** [comment]
I'm sorry @raghavsethi I didn't get time to attend this in a more timely manner.

**Comment by raghavsethi** [comment]
@ilfrin I discussed this PR with @martint and we believe that this approach is problematic if the only action you want to take is add a stat for total spilled bytes. I believe you've tried to emulate the memory tracking, but memory is specifically reserved and freed, and most of the tooling exists to make that possible. For spill, you simply need an operator stat that is rolled up properly all the way to the query level.

@martint can chime in with other ideas.

**Comment by kokosing** [comment]
@ilfrin I think we could close this, don't you think?

**Comment by raghavsethi** [comment]
Agreed, let's close it. There's too many conflicts at this point and the approach is problematic.
